src/main/scala/org/opennetworkinsight/dns/DNSWordCreation.scala

```diff
@@ -1,10 +1,135 @@
 package org.opennetworkinsight.dns
 
+import org.apache.spark.SparkContext
 import org.apache.spark.broadcast.Broadcast
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types.{DoubleType, StringType, StructField, StructType}
+import org.opennetworkinsight.utilities.{Entropy, Quantiles}
+import org.opennetworkinsight.dns.{DNSSchema => Schema}
+import org.slf4j.Logger
+
+import scala.io.Source
 
   object DNSWordCreation {
 
-    val l_country_codes = Set("ac", "ad", "ae", "af", "ag", "ai", "al", "am", "an", "ao", "aq", "ar", "as", "at", "au",
+    def dnsWordCreation(totalDataDF: DataFrame,
+                        sc: SparkContext, logger: Logger, sqlContext: SQLContext): DataFrame ={
+
+      var timeCuts = new Array[Double](10)
+      var frameLengthCuts = new Array[Double](10)
+      var subdomainLengthCuts = new Array[Double](5)
+      var numberPeriodsCuts = new Array[Double](5)
+      var entropyCuts = new Array[Double](5)
+
+      val topDomains = sc.broadcast(Source.fromFile("top-1m.csv").getLines.map(line => {
+        val parts = line.split(",")
+        parts(1).split("[.]")(0)
+      }).toSet)
+
+      val countryCodesBC = sc.broadcast(countryCodes)
+
+      logger.info("Computing subdomain info")
+
+      val queryNameIndex = totalDataDF.schema.fieldNames.indexOf(Schema.QueryName)
+
+      val dataWithSubdomainsRDD: RDD[Row] = totalDataDF.rdd.map(row =>
+        Row.fromSeq {
+          row.toSeq ++
+            extractSubdomain(countryCodesBC, row.getString(queryNameIndex))
+        })
+
+      // Update data frame schema with newly added columns. This happens b/c we are adding more than one column at once.
+      val schemaWithSubdomain = {
+        StructType(totalDataDF.schema.fields ++
+          refArrayOps(Array(StructField(Schema.Domain, StringType),
+            StructField(Schema.Subdomain, StringType),
+            StructField(Schema.SubdomainLength, DoubleType),
+            StructField(Schema.NumPeriods, DoubleType))))
+      }
+
+      val dataWithSubDomainsDF = sqlContext.createDataFrame(dataWithSubdomainsRDD, schemaWithSubdomain)
+
+      logger.info("Computing subdomain entropy")
+
+      val udfStringEntropy = DNSWordCreation.udfStringEntropy()
+
+      val dataWithSubdomainEntropyDF = dataWithSubDomainsDF.withColumn(Schema.SubdomainEntropy,
+        udfStringEntropy(col(Schema.Subdomain)))
+
+      logger.info("Calculating time cuts ...")
+
+      timeCuts = Quantiles.computeDeciles(dataWithSubdomainEntropyDF
```

Comment:
this isn't quite the same structure as is used in the proxy code...

with proxy,  the calculation of the quantiles and their cuts is placed inside the preLDA code, at that level
whereas for DNS,  the cut calculations are placed inside the word creation

I think that placing the quantile/cut calculations at the higher level makes sense because eventually we will want to move them to a level where the cuts are calculated up front and then passed to both the model building (preLDA) and scoring (postLDA) phases, so that postLDA doesn't have to recalculate the things

but for now, just get the statistical calculation separated from the word calculation (and the logic looking pretty much the same as that of proxy)